src/core.jl

```diff
@@ -239,33 +64,44 @@ Returns a `StatsBase.Histogram` of the degrees of vertices in `g`.
 """
 degree_histogram(g::AbstractGraph) = fit(Histogram, degree(g))
 
-"""Returns a list of all neighbors connected to vertex `v` by an incoming edge.
-
-NOTE: returns a reference, not a copy. Do not modify result.
 """
-in_neighbors(g::AbstractGraph, v::Int) = badj(g,v)
-"""Returns a list of all neighbors connected to vertex `v` by an outgoing edge.
-
+Return a list of all neighbors reachable from vertex `v` in `g`.
+For DiGraphs, the default is equivalent to `out_neighbors(g, v)`;
+use `all_neighbors` to list inbound and outbound neighbors.
 NOTE: returns a reference, not a copy. Do not modify result.
 """
-out_neighbors(g::AbstractGraph, v::Int) = fadj(g,v)
+neighbors(g::AbstractGraph, v::Integer) = out_neighbors(g, v)
 
-"""Returns a list of all neighbors of vertex `v` in `g`.
-
-For DiGraphs, this is equivalent to `out_neighbors(g, v)`.
-
-NOTE: returns a reference, not a copy. Do not modify result.
 """
-neighbors(g::AbstractGraph, v::Int) = out_neighbors(g, v)
+Return a list of all inbound and outbount neighbors of `v` in `g`.
+For undirected graphs, this is equivalent to `out_neighbors` and
+`in_neighbors`.
+"""
+all_neighbors(x...) = _NI("all_neighbors")
+@traitfn all_neighbors{G<:AbstractGraph; IsDirected{G}}(g::G, v::Integer) =
+  union(out_neighbors(g, v), in_neighbors(g, v))
+@traitfn all_neighbors{G<:AbstractGraph; !IsDirected{G}}(g::G, v::Integer) =
+  neighbors(g, v)
 
-"Returns the neighbors common to vertices `u` and `v` in `g`."
-common_neighbors(g::AbstractGraph, u::Int, v::Int) = intersect(neighbors(g,u), neighbors(g,v))
 
-@deprecate has_self_loop has_self_loops
+"Returns the neighbors common to vertices `u` and `v` in `g`."
+common_neighbors(g::AbstractGraph, u::Integer, v::Integer) =
+  intersect(neighbors(g, u), neighbors(g, v))
 
 "Returns true if `g` has any self loops."
-
-has_self_loops(g::AbstractGraph) = any(v->has_edge(g, v, v), vertices(g))
+has_self_loops(g::AbstractGraph) = nv(g) == 0? false : any(v->has_edge(g, v, v), vertices(g))
 
 "Returns the number of self loops in `g`."
-num_self_loops(g::AbstractGraph) = sum(v->has_edge(g, v, v), vertices(g))
+num_self_loops(g::AbstractGraph) = nv(g) == 0 ? 0 : sum(v->has_edge(g, v, v), vertices(g))
```

Comment:
If a type supports self loops you might need to do something different even if there are no self loops present. I am thinking about having to allocate spots for the self loops even if they aren't there.